hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java

```diff
@@ -1008,12 +1012,56 @@ private TransactionCompletionProcesses(
 		 */
 		public static final InsertActionSorter INSTANCE = new InsertActionSorter();
 
+		private static class BatchIdentifier {
+
+			private final String entityName;
+
+			private List<String> parentEntityNames = new ArrayList<>(  );
```

Comment:
IMO both `parentEntityNames` and `childEntityNames` could be HashSet<String>.  Strings have very efficient hashCode impls and considering you later perform explicit `List#contains` checks, I think the Set-approach would be faster.  Unless I miss some need for the order.